lib/jobs.go

```diff
@@ -0,0 +1,117 @@
+package lib
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"github.com/rs/xid"
+	log "github.com/sirupsen/logrus"
+)
+
+// allJobs stores the details of all jobs in memory.
+// It maps Job object using the job ID for faster lookup.
+var allJobs = make(map[string]*Job)
+
+// CreateJob takes a CreateJobRequest, starts the job, and returns the Job object.
+// It returns error if job fails to start.
+// On success, it creates a new Job object, assign a unique Job.ID, and sets Job.Status to JobRunning.
+// It stores the job in memory against the Job.ID.
+// It creates a new go routine that waits for job to exit.
+func CreateJob(request CreateJobRequest) (Job, error) {
+	// combines stderr and stdout of the job process
+	var outputBuffer bytes.Buffer
+
+	cmd := exec.Command(request.Command, request.Args...)
+	cmd.Stdout = &outputBuffer
+	cmd.Stderr = &outputBuffer
+
+	// no need to save the job if it fails to start
+	if err := cmd.Start(); err != nil {
+		err = fmt.Errorf("job failed to start, reason: %v", err)
+		log.Errorf(err.Error())
+		return Job{}, err
+	}
+
+	// job started, create job object, assign ID and save the job
+	job := Job{
+		ID:           getNextJobID(),
+		Request:      request,
+		Status:       JobRunning,
+		cmd:          cmd,
+		outputBuffer: &outputBuffer,
+	}
+
+	log.Infof("job %v: started", job.ID)
+
+	// save the job in memory for future reference
+	allJobs[job.ID] = &job
+
+	// wait for job to finish
+	go handleFinish(&job)
+
+	return job, nil
+}
+
+// StopJob takes a job ID, fetch the job from memory, and kills the job process.
+// It returns error if job is not found or if job can not be terminated.
+// On success, it waits for cmd.Wait to finish which sets the Job.Status to JobExited.
+func StopJob(id string) error {
+	job, ok := allJobs[id]
+	if !ok {
+		return fmt.Errorf("job %v: not found", id)
+	}
+	if job.Status == JobExited {
```

Comment:
This is a race condition against the write to `job.Status` in `handleFinish`

Reply:
@diptadas this race condition is still present, since you're modifying the same `*Job` pointer

Reply:
Added RW lock for job status